contracts/src/registry/L1ETHRegistry.sol

```diff
@@ -0,0 +1,235 @@
+    mapping(uint256 => address) public tokenObservers;
+
+    address public ejectionController;
+
+    constructor(IRegistryDatastore _datastore, address _ejectionController) PermissionedRegistry(_datastore) {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        // Grant role to the ejection controller
+        require(_ejectionController != address(0), "Ejection controller cannot be empty");
+
+        ejectionController = _ejectionController;
+        _grantRole(EJECTION_CONTROLLER_ROLE, _ejectionController);
+    }
+
+    function uri(uint256 /*tokenId*/ ) public pure override returns (string memory) {
+        // TODO: implement metadata uri
+        return "";
+    }
+
+    function ownerOf(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC1155Singleton, IERC1155Singleton)
+        returns (address)
+    {
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 expires = _extractExpiry(oldFlags);
+        if (expires < block.timestamp) {
+            return address(0);
+        }
+        return super.ownerOf(tokenId);
+    }
+
+    /**
+     * @dev Receive an ejected name from L2.
+     * @param labelHash The keccak256 hash of the label
+     * @param owner The owner of the name
+     * @param registry The registry to use for the name
+     * @param flags The base flags
+     * @param expires Expiration timestamp
+     * @return tokenId The token ID of the ejected name
+     */
+    function ejectFromL2(uint256 labelHash, address owner, IRegistry registry, uint32 flags, uint64 expires)
+        public
+        onlyRole(EJECTION_CONTROLLER_ROLE)
+        returns (uint256 tokenId)
+    {
+        tokenId = (labelHash & ~uint256(FLAGS_MASK)) | flags;
+        uint96 fullFlags = (uint96(flags) & FLAGS_MASK) | (uint96(expires) << 32);
+
+        // Check if there is a previous owner and verify the name is expired
+        // to prevent malicious ejection controllers from overriding valid L1 names
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+            uint64 oldExpires = _extractExpiry(oldFlags);
+            if (oldExpires >= block.timestamp) {
+                revert NameNotExpired(tokenId, oldExpires);
+            }
+            _burn(previousOwner, tokenId, 1);
+        }
+
+        _mint(owner, tokenId, 1, "");
+        datastore.setSubregistry(tokenId, address(registry), fullFlags);
+        emit NameEjected(tokenId, owner, expires);
+        return tokenId;
+    }
+
+    /**
+     * @dev Renew an ejected name.
+     * After renewal, the ejection controller should be notified to sync the update to L2.
+     *
+     * @param tokenId The token ID of the name to renew
+     * @param expires New expiration timestamp
+     */
+    function renew(uint256 tokenId, uint64 expires) public {
```

Comment:
This lets users renew names indefinitely for free, which seems bad. We should add a controller role here and only allow that contract to do renewals.